Single target testing

ppenzin opened the pull request:
Enable building and testing via a single target (check) - require chakracore
library to be built before the test run.

For #6547

ppenzin committed:
Single target testing

Enable building and testing via a single target (check) - require chakracore
library to be built before the test run.

ppenzin commented:
The idea is that we can invoke one target, which would be `check` for debug builds, and default for release builds, this would keep CI build steps for all builds mostly identical.

ppenzin committed:
[CI] Enable Linux release builds

Manipulate build target: run tests for debug builds and simply build for release.

ppenzin committed:
Add OSX CI

Expand CI matrix to include OSX.

ppenzin committed:
Typo - duplicate yaml tag

ppenzin commented:
Three Linux builds are working, trying to add OSX. I am going to make sure all basic varieties work, then will try to replicate some of the testing old CI used to do. I am not going to wait for this - will check tomorrow.

With this much variety we are getting close to the number of jobs Azure allows for free - after 10 jobs are launched the rest would have to wait their turn.

rhuanjl commented:
The macOS debug build is failing all the tests but rather than saying it's failed it's carrying on failing over and over.

Two problems to solve:
1. Why is it failing everything (some failures look like missing ICU but there are others which that doesn't explain)?
2. Why is it carrying on when it's failed? Is there anything we can change/set to make it stop when it fails?

EDIT: the test failures are this issue: #5876 it's a long standing problem with PAL that surfaces in macOS shared debug builds - the old CI dodged it by using a static build.

ppenzin commented:
The test runner tries to run until the end. I will set up a timeout - I was hoping Azure would kill it, but it did not.

> EDIT: the test failures are this issue: #5876 it's a long standing problem with PAL that surfaces in macOS shared debug builds - the old CI dodged it by using a static build.

Thank you for finding it! Converting the build to static should be a relatively easy fix.

rhuanjl commented:
> The test runner tries to run until the end. I will set up a timeout - I was hoping Azure would kill it, but it did not.
> 
> > EDIT: the test failures are this issue: #5876 it's a long standing problem with PAL that surfaces in macOS shared debug builds - the old CI dodged it by using a static build.
> 
> Thank you for finding it! Converting the build to static should be a relatively easy fix.

I wonder if we could/should add a "terminate on failure" flag to runtests.py or even make that the default behaviour with a flag to cancel it if you want to find multiple failures offline. I could have a look at how to do that.

rhuanjl commented on azure-pipelines.yml:
```diff
@@ -3,19 +3,46 @@ trigger:
 jobs:
-  - job: Linux
+  - job: CMake
     timeoutInMinutes: 0
```
Should we do a 2 hour time out (per job)?

ppenzin replied:
Yep, just did that.

rhuanjl commented on azure-pipelines.yml:
```diff
@@ -3,19 +3,46 @@ trigger:
+          test_target: 'all'
+        OSX.Debug:
+          image_name: 'macOS-latest'
+          deps: 'brew install ninja icu4c'
```
Can we add a flag to each config for static vs shared?

(And obviously need to set OSX.Debug to static so its tests will pass)

ppenzin replied:
Yep, that is in the new commit.

ppenzin committed:
Timeout and OSX debug library type

Prevent tests from failing on OSX debug builds, by setting timeouts and
forcing static library build.

Shared library issue is described in #5876.

ppenzin commented:
> I wonder if we could/should add a "terminate on failure" flag to runtests.py or even make that the default behaviour with a flag to cancel it if you want to find multiple failures offline. I could have a look at how to do that.

Maybe - I think it is useful to see all failures, though we can fail on the first one if it is too expensive to keep running. I wondering why the 'nix testing is slower - maybe we can improve the script somehow.

ppenzin commented:
This does not seem to help, not sure what I am doing wrong.

ppenzin committed:
CI debug: library flag

ppenzin committed:
Use proper shell syntax

Fix environment variable reference. Remove debug output.

ppenzin committed:
Merge branch 'master' into ci-single-target

ppenzin commented:
Turns out I had brackets around an environment variable name :)
